### PR TITLE
Torch motion

### PR DESCRIPTION
**Pull request by GReguig**
Hello,

I added a generic way to compute metrics after a transform and also modified the RandomMotionFromTimeCourse in order to avoid the calculation of the metrics inside the transform

### CODE REVIEW

File: torchio/metrics/metric_wrapper.py [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+            if sample_key is self.mask_key:
+                continue
+            data1 = sample1[sample_key][DATA]
+            data2 = sample2[sample_key][DATA]
+
+            if self.use_mask and self.mask_key is not None:
+                mask_data1, mask_data2 = sample1[self.mask_key][DATA], sample2[self.mask_key][DATA]
+                data1 = torch.mul(data1, mask_data1)
+                data2 = torch.mul(data2, mask_data2)
+            #Apply mask if mask
```
Comment (romainVala) [inline torchio/metrics/metric_wrapper.py]: J'ai l'impression que la tu applique le mask avant de calculer la distance entre les 2 images.
dans l'implementation
https://github.com/romainVala/torchio/blob/11377907587d2ff39f5a62ef4b9d15aaa1b7934a/torchio/transforms/augmentation/intensity/random_motion_from_time_course.py#L172
le mask key peut etre une liste (par ex: brain, gray matter, white matter ...)
dans ce cas il faut calculer un "metric_map", sur toute l'image, est ensuite c'est au moment de faire la moyenne qu'il faut boucler sur les mask
donc la metric revoit une list de metric (de meme taille que la liste de mask)

File: torchio/metrics/metric_wrapper.py [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+            """
+            if self.use_mask and self.mask_key is not None:
+                mask_data = sample[self.mask_key][DATA]
+                orig_data = torch.mul(orig_data, mask_data)
+                transformed_data = torch.mul(transformed_data, mask_data)
+            """
+            #Compute metric
+            if "metrics" not in sample2[sample_key].keys():
+                sample2[sample_key]["metrics"] = dict()
+            sample2[sample_key]["metrics"][self.metric_name] = self.metric_func(data1, data2)
```
Comment (romainVala) [inline torchio/metrics/metric_wrapper.py:30]: pour continuer le commentaire d'avant, (sur les mask)
On peut le gerer ici, et du coup metric_func, ne renvoie pas un nombre mais une map
a partir de cette map on fait soit la moyenne sur toute l'image soit la moyenne sur les mask. Le nom du résultat devient self.metric_name+ '_' + self.mask_key

File: torchio/transforms/augmentation/intensity/random_motion_from_time_course.py [hunk 3]
```diff
@@ -405,10 +373,10 @@ def _simulate_random_trajectory(self):
...
 
         self.fitpars = fitpars
-        #print(f' in _simul_motionfitpar shape fitpars {fitpars.shape}')
-        simu_param = dict(noisPar=noiseBasePars,maxDisp=maxDisp,maxRot=maxRot,
-                          swallowFrequency=swallowFrequency, swallowMagnitudeT=swallowMagnitude[0], swallowMagnitudeR=swallowMagnitude[1],
-                          suddenFrequency=suddenFrequency,suddenMagnitudeT=suddenMagnitude[0], suddenMagnitude=suddenMagnitude[1])
+        simu_param = dict(fitpars=fitpars, noisPar=noiseBasePars, maxDisp=maxDisp, maxRot=maxRot,
+                          swallowFrequency=swallowFrequency, swallowMagnitudeT=swallowMagnitude[0],
+                          swallowMagnitudeR=swallowMagnitude[1], suddenFrequency=suddenFrequency,
+                          suddenMagnitudeT=suddenMagnitude[0], suddenMagnitude=suddenMagnitude[1])
```
Comment (romainVala) [inline torchio/transforms/augmentation/intensity/random_motion_from_time_course.py:372]: c'est ici qu'il faut ajouter les metric sur les parametres de mouvement

File: torchio/transforms/augmentation/intensity/random_motion_from_time_course.py [hunk 4]
```diff
@@ -158,53 +155,24 @@ def apply_transform(self, sample):
...
+        metrics['mean_DispP_iterp'] = calculate_mean_Disp_P(fitpars_interp)
+        metrics['rmse_Disp_iterp'] = calculate_mean_RMSE_displacment(fitpars_interp)
+
+        ff_interp, to_substract = self.demean_fitpar(fitpars_interp, original_image)
+        metrics['TFsubstract'] = to_substract
+        metrics['rmse_DispTF'] = calculate_mean_RMSE_displacment(ff_interp, original_image)
+
+        parameters_motion['metrics_motion'] = metrics
+
+        sample.add_transform(self, parameters_motion)
```
Comment (romainVala) [inline torchio/transforms/augmentation/intensity/random_motion_from_time_course.py:163]: les metriques sur les fitpar, peuvent etre deplace plus loin (a la fin de simu_motion)

File: torchio/transforms/augmentation/composition.py [hunk 5]
```diff
@@ -20,8 +20,8 @@ class Compose(Transform):
     .. note::
         This is a thin wrapper of :py:class:`torchvision.transforms.Compose`.
     """
-    def __init__(self, transforms: Sequence[Transform], p: float = 1):
-        super().__init__(p=p)
+    def __init__(self, transforms: Sequence[Transform], p: float = 1, **kwargs):
+        super().__init__(p=p, **kwargs)
```
Comment (romainVala) [inline torchio/transforms/augmentation/composition.py:24]: Ok ici il faut le garder

File: torchio/transforms/augmentation/intensity/random_swap.py [hunk 6]
```diff
@@ -24,8 +24,9 @@ def __init__(
             num_iterations: int = 100,
             p: float = 1,
             seed: Optional[int] = None,
+            **kwargs
             ):
-        super().__init__(p=p, seed=seed)
+        super().__init__(p=p, seed=seed, **kwargs)
```
Comment (romainVala) [inline torchio/transforms/augmentation/intensity/random_swap.py:29]: pour cette transfo, c'est peut etre pas la peine de l'ajouter (meme si ca coute rien) car ca a pas trop de sens de vouloir caluler une distance apres un swap ...

File: torchio/transforms/augmentation/intensity/torch_random_motion.py [hunk 7]
```diff
@@ -151,48 +147,23 @@ def apply_transform(self, sample):
...
-                sample[image_name]['metrics'] = metrics
+        ff_interp, to_substract = self.demean_fitpar(fitpars_interp, original_image)
+        metrics['TFsubstract'] = to_substract
+        metrics['rmse_DispTF'] = calculate_mean_RMSE_displacment(ff_interp, original_image)
 
-            if self.res_dir is not None:
-                self.save_to_dir(image_dict)
+        parameters_motion['metrics_motion'] = metrics
 
+        sample.add_transform(self, parameters_motion)
```
Comment (romainVala) [inline torchio/transforms/augmentation/intensity/torch_random_motion.py:166]: comme pour random_motion_from_time_course, a mettre dans la fonction simu_param

File: torchio/transforms/preprocessing/intensity/apply_mask.py [hunk 8]
```diff
@@ -15,8 +15,9 @@ def __init__(
             self,
             masking_method: Union[str, TypeCallable, None] = None,
             p: float = 1,
+            **kwargs
             ):
-        super().__init__(masking_method=masking_method, p=p)
+        super().__init__(masking_method=masking_method, p=p, **kwargs)
```
Comment (romainVala) [inline torchio/transforms/preprocessing/intensity/apply_mask.py:21]: peut etre pas la peine ici non plus

File: examples/tests_metrics.py [hunk 9]
```diff
@@ -0,0 +1,33 @@
...
+mask_path = "/data/romain/HCPdata/suj_100307/cat12/fill_mask_head.nii.gz"
+dataset = ImagesDataset([
+    Subject({
+        "T1": Image(t1_path),
+        "mask": Image(mask_path)
+    })])
+
+metrics = {
+    "L1": MetricWrapper("L1", L1Loss()),
+    "L2": MetricWrapper("L2", MSELoss()),
```
Comment (romainVala) [inline examples/tests_metrics.py:21]: les loss renvoient une moyenne de la distance, (Cf commentaire d'apres) il faut plutot utiliser L1 (pour renvoyer une map). 
Et dans le metricWrapper on gère le mean

File: examples/tests_metrics.py [hunk 10]
```diff
@@ -0,0 +1,33 @@
...
+    "SSIM": SSIM3D(return_map=False, discard_zeros=True),
+    "SSIM_Wrapped": MetricWrapper("SSIM_wrapped", lambda x, y: functional_ssim(x, y, discard_zeros=True), use_mask=True, mask_key="mask"),
```
Comment (romainVala) [inline examples/tests_metrics.py]: il faudrait que dans ton exemple tu prenne une list de mask
ajoute dans le Subject p1 p2
et du coup mask_key = ["mask", "p1", "p2"]

File: torchio/metrics/ssim.py [hunk 11]
```diff
@@ -0,0 +1,122 @@
...
+
+    if not return_map:
+
+        if discard_zeros:
+            non_zero_vals = torch.mul((x > 0.), (y > 0.))
+            luminance = luminance[non_zero_vals]
+            contrast = contrast[non_zero_vals]
+            structure = structure[non_zero_vals]
+            ssim = ssim[non_zero_vals]
+
```
Comment (romainVala) [inline torchio/metrics/ssim.py]: il faut enlever ce parametre discard_zeros, (pour gérer avec les masks, la moyenne des metriques maps, dans le MetricWrapper

**Review by romainVala — COMMENTED** [review]
Merci Ghiles, 
c'est un chouette changement de pouvoir l'utiliser facilement sur toutes les transfo (sans avoir besoin du keep original)

Pour moi il manque essentiellement la gestion des masks:
il ne faut pas appliquer les mask aux data avant le calcule de la metrique, mais plustot avoir des metriques qui renvoit des map (volume) et appliquer les mask sur cette map
ca donne pas le meme résulatat, et en plus on peut facilement avoir un liste de mask

File: torchio/transforms/augmentation/intensity/random_swap.py (hunk 6 above)
Comment (GReguig) [reply torchio/transforms/augmentation/intensity/random_swap.py:29]: C'est plus pour que ça soit général (si jamais on ajoute un autre argulent à la classe transform de base, on retrouvera cet argument directement dans toutes les transforms, ce n'est pas fait spécifiquement pour le calcul des métriques)

File: torchio/metrics/metric_wrapper.py (hunk 1 above)
Comment (GReguig) [reply torchio/metrics/metric_wrapper.py]: Je pensais laisser ce genre d'option (application du mask après calcul de la métrique) pour les métriques en question. Mais effectivement, je vais modifier ça pour que ça corresponde à ta proposition

File: torchio/transforms/augmentation/intensity/random_swap.py (hunk 6 above)
Comment (romainVala) [reply torchio/transforms/augmentation/intensity/random_swap.py:29]: OK

**Comment by GReguig** [comment]
Je viens de push une nouvelle version, si tu as le temps d'en faire une review. Je t'avoue que je ne suis pas complétement satisfait de la structure du code (à quel moment les calculs des masques des métriques sont faits), mais je ne vois pas trop comment faire autrement

File: torchio/metrics/map_metric.py [hunk 12]
```diff
@@ -0,0 +1,47 @@
...
+        if self.mask_keys:
+            mask_keys_in_sample = sample.keys() & set(self.mask_keys)
+
+            if len(mask_keys_in_sample) == 0:
+                warnings.warn("None of the given masks {} found for the sample".format(self.mask_keys))
+            else:
+
+                for mask_key in mask_keys_in_sample:
+                    mask_data = sample[mask_key][DATA]
+                    masked_map = torch.mul(mask_data, metric_map)
```
Comment (romainVala) [inline torchio/metrics/map_metric.py]: tu prend bien tous les mask, mais en sortie masked_map ne prend que la derniere valeur de la boucle ?

File: torchio/metrics/map_metric.py [hunk 13]
```diff
@@ -0,0 +1,47 @@
...
+
+        if self.average_method is not None:
+            masked_map = self.average_method(masked_map)
+        return masked_map
```
Comment (romainVala) [inline torchio/metrics/map_metric.py]: la gestion des mask n'est toujours pas bonne (je crois), car tu fais la moyenne en prennant en compte tous les zeros ..
toi tu fais :
metrics_result = torch.mean( torch.mul(mask_data, metric_map))
si metric_map = [2, 2, 2, 2] et que le mask=[0, 0, 1,1]
ca va faire mean([0,0,2,2]) = 1
moi je veux mean([2,2]) = 2

donc il faut faire 

metrics_result = torch.mean(metric_map[mask_data>0] ) 
et avec une list de mask
metrics_result = [ torch.mean(metric_masp[mm>0] for mm in mask_data_list] 

ensuite il faut construire un dictionaire avec les champs :  nom de la metric + '_' + mask_name

sinon tu peux garde the self.average_method, si tu change la boucle for 41 -> 43 par
masked_map = [ metric_map[sample[mask_key][DATA] ]  for mask_key in mask_keys_in_sample ]

il reste juste a construire le dic de resultat

c'est clair ?

File: torchio/metrics/map_metric_wrapper.py [hunk 14]
```diff
@@ -0,0 +1,26 @@
...
+                continue
+            data1 = sample1[sample_key][DATA]
+            data2 = sample2[sample_key][DATA]
+
+            if "metrics" not in sample2[sample_key].keys():
+                sample2[sample_key]["metrics"] = dict()
+            metric_map = self.metric_func(data1, data2)
+
+            metric_map = self._apply_masks_and_averaging(sample2, metric_map=metric_map)
+            sample2[sample_key]["metrics"][self.metric_name] = metric_map
```
Comment (romainVala) [inline torchio/metrics/map_metric_wrapper.py:26]: peut etre que c'est plus simple de faire la boucle sur les differents mask, ici ...
et comme ca tu peux faire la concatenation :  
sample2[sample_key]["metrics"][self.metric_name+ '_' + mask_key] =

**Review by romainVala — COMMENTED** [review]
on y est presque

**Comment by GReguig** [comment]
Je pense que maintenant, ça doit être bon. J'ai implémenté l'utilisation de plusieurs masques pour les métriques. Aussi, on renvoit aussi systèmatiquement le résultat sans aucun masque (avec le tag "no_mask")

**Comment by romainVala** [comment]
Merci Ghiles !
